### ISSUE TITLE
PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 timing-out/hanging in ATDM Trilihnos CUDA opt builds starting 2020-02-02

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#6801]
CC: @trilinos/panzer, @mperego (Trilinos Discretizations Product Lead)

<Checklist>
<???: Add label "client: ATDM">
<???: Add label "ATDM Sev: Blocker" (by default but could be other "ATDM Sev: XXX")>
<???: Add label "type: bug"?>
<???: Add label for affected packages (e.g. "pkg: MueLu", "pkg: Tpetra", "pkg: Kokkos", etc.)>
<???: Add label "PA: ???Project Area???" (e.g. "PA: Linear Solvers", "PA: Data Services")>
<???: Add milestone "Initial cleanup of new ATDM ..." or "Keep promoted ATDM ...">
<???: Once GitHub Issue is created, add entries for tests to TrilinosATDMStatus/*.csv files>

## Next Action Status

<status-and-or-first-action>

## Description

As shown in [this query](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-01-29&end=2020-02-09&filtercount=3&showfilters=1&filtercombine=and&field1=block&field1count=2&field1field1=buildname&field1compare1=61&field1value1=Trilinos-atdm-waterman-cuda-9.2-opt&field1field2=buildname&field1compare2=61&field1value2=Trilinos-atdm-waterman_cuda-9.2_fpic_static_opt&field2=testname&compare2=61&value2=PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1&field3=site&compare3=61&value3=waterman) the test:

* `PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1`

in the builds:

* `Trilinos-atdm-waterman-cuda-9.2-opt`
* `Trilinos-atdm-waterman_cuda-9.2_fpic_static_opt`

started timing out on testing day 2020-02-02.

Before they starting timing out at 10 minutes, these tests were finishing in under 4 minutes, some days in under 1.5 minutes.  Also, as shown [for all builds on 'waterman' on 2020-02-09](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&date=2020-02-09&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-waterman&field2=testname&compare2=61&value2=PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1&field3=site&compare3=61&value3=waterman):

Build Name | Test Name | Status | Time | Details | Build Time
-- | -- | -- | -- | -- | --
Trilinos-atdm-waterman-cuda-9.2-opt | PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 | Failed | 10m 110ms | Completed (Timeout) | 2020-02-09T13:02:16 MST
Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug | PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 | Passed | 1m 52s 160ms | Completed | 2020-02-09T03:27:44 MST
Trilinos-atdm-waterman-cuda-9.2-release-debug | PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 | Passed | 1m 55s 520ms | Completed | 2020-02-09T13:02:35 MST
Trilinos-atdm-waterman_cuda-9.2_fpic_static_opt | PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 | Failed | 10m 150ms | Completed (Timeout) | 2020-02-09T10:06:38 MST
Trilinos-atdm-waterman_cuda-9.2_shared_opt | PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 | Passed | 9m 1s 230ms | Completed | 2020-02-09T08:03:56 MST

this test passed in less than 1 minutes in debug builds.  This suggests that this test is hanging and is not just running slowly.

## Current Status on CDash

* [PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 test on 'waterman' builds current testing day](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-waterman&field2=testname&compare2=61&value2=PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1&field3=site&compare3=61&value3=waterman)
* [PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 test in these 2 'waterman' builds last 5 days](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=5%20days%20ago&end=now&filtercount=3&showfilters=1&filtercombine=and&field1=block&field1count=2&field1field1=buildname&field1compare1=61&field1value1=Trilinos-atdm-waterman-cuda-9.2-opt&field1field2=buildname&field1compare2=61&field1value2=Trilinos-atdm-waterman_cuda-9.2_fpic_static_opt&field2=testname&compare2=61&value2=PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1&field3=site&compare3=61&value3=waterman)

## Steps to Reproduce

One should be able to reproduce this failure on the machine 'waterman' as described in:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md

More specifically, the commands given for the system 'waterman' are provided at:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md#waterman

The exact commands to reproduce this hanging test for the build `Trilinos-atdm-waterman-cuda-9.2-opt`, for example, should be:

```
$ cd <some_build_dir>/

$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh \
    Trilinos-atdm-waterman-cuda-9.2-opt

$ cmake \
 -GNinja \
 -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
 -DTrilinos_ENABLE_TESTS=ON -DTrilinos_ENABLE_Panzer=ON \
 $TRILINOS_DIR

$ make NP=16

$ bsub -x -Is -n 20 ctest -R PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 
```

**Comment by rppawlo** [comment]
Having trouble reproducing:
```
bash-4.2$ ctest -R PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1
Test project /ascldap/users/rppawlo/Trilinos2/BUILD_OPT/packages/panzer/adapters-stk/example/PoissonInterfaceExample
    Start 2: PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1
1/1 Test #2: PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 ...   Passed   59.87 sec
```

```
bash-4.2$ ctest -j4
Test project /ascldap/users/rppawlo/Trilinos2/BUILD_OPT/packages/panzer/adapters-stk/example/PoissonInterfaceExample
    Start 3: PanzerAdaptersSTK_PoissonInterfaceExample_3d_MPI_4
1/5 Test #3: PanzerAdaptersSTK_PoissonInterfaceExample_3d_MPI_4 ...............   Passed   25.14 sec
    Start 1: PanzerAdaptersSTK_PoissonInterfaceExample_2d_MPI_4
2/5 Test #1: PanzerAdaptersSTK_PoissonInterfaceExample_2d_MPI_4 ...............   Passed   18.65 sec
    Start 2: PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1
3/5 Test #2: PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 ...   Passed   59.71 sec
    Start 5: PanzerAdaptersSTK_PoissonInterfaceExample_3d_MPI_4-ConvTest
4/5 Test #5: PanzerAdaptersSTK_PoissonInterfaceExample_3d_MPI_4-ConvTest ......   Passed    8.76 sec
    Start 4: PanzerAdaptersSTK_PoissonInterfaceExample_2d_MPI_4-ConvTest
5/5 Test #4: PanzerAdaptersSTK_PoissonInterfaceExample_2d_MPI_4-ConvTest ......   Passed    6.23 sec

100% tests passed, 0 tests failed out of 5

Label Time Summary:
Panzer    = 294.86 sec*proc (5 tests)

Total Test time (real) = 118.51 sec
bash-4.2$ 
```
could this be an artifact of overloading the gpu? This is in a cuda opt build.

**Comment by bartlettroscoe** [comment]
@rppawlo, interesting, this must be due then to contention for the GPU even from just 4 one-process MPI tests running at the same time completing for the same GPU.   The reproducer might then need to be:

```
$ cd <some_build_dir>/

$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh \
    Trilinos-atdm-waterman-cuda-9.2-opt

$ cmake \
 -GNinja \
 -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
 -DTrilinos_ENABLE_TESTS=ON -DTrilinos_ENABLE_ALL_PACKAGES=ON \
 $TRILINOS_DIR

$ make NP=16

$ bsub -x -Is -n 20 ctest -j 4
```

Wonder what has changed to cause such problems with sharing the GPU?

What is crazy about this is that if you look at the performance of this test across all of the build CUDA builds on 'waterman' in [this query](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-01-29&end=2020-02-09&filtercount=4&showfilters=1&filtercombine=and&field1=groupname&compare1=62&value1=Experimental&field2=buildname&compare2=65&value2=Trilinos-atdm-waterman&field3=buildname&compare3=63&value3=cuda&field4=testname&compare4=61&value4=PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1) you see that the test run and finishes just find in the builds:

* `Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug` (less than 3m)
* `Trilinos-atdm-waterman-cuda-9.2-release-debug` (less than 3m)

How can this test be running so fast in debug builds but be having such problems running in optimized builds when run with other tests on the GPU?  But we did seem similar problems with the pure optimized builds or with just the shared or -fPIC builds in issues #6805 and #6804.

Kitware is supposed to be helping us with this as per https://gitlab.kitware.com/snl/project-1/issues/100.   Otherwise, we can drop all of the way down to `ctest -j1` for these optimized buiolds but we will have problems getting all of our builds completed in a single testing day.

What has changed on the system or in the code to have caused this change in behavior with respect to running multiple jobs on the same GPU no 'waterman'?  We don't seem to be seeing a problem on 'ride' as shown in [this query](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-01-29&end=2020-02-09&filtercount=4&showfilters=1&filtercombine=and&field1=groupname&compare1=62&value1=Experimental&field2=buildname&compare2=65&value2=Trilinos-atdm-white-ride&field3=buildname&compare3=63&value3=cuda&field4=testname&compare4=61&value4=PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1).

**Comment by mhoemmen** [comment]
Are we running the MPS server there?

**Comment by rppawlo** [comment]
> Are we running the MPS server there?

I don't think so. I ran the check from the mps documentation and it came back negative.

**Comment by bartlettroscoe** [comment]
> Are we running the MPS server there?

No, but that is something we could try if it is supported on 'waterman'.  Where is the documentation for that?

**Comment by rppawlo** [comment]
@bartlettroscoe - I followed the steps in your last post. The first run through of the entire test suite had only one test failing (hard failure in muelu plist interpreter). On the second run through, the tests were reordered since ctest stored the runtimes and uses to optimize testing time.  This second time through I got all the timeouts below. I think all the slow single process gpu intensive tests get launched together at the very beginning of a retest and jam the gpu. I wonder if disabling the ctest optimization might help with this? Deleting the test output directory to wipe the timing info eliminates all the timeouts on a rerun.

```
Test project /ascldap/users/rppawlo/Trilinos2/BUILD_OPT
        Start 1621: MueLu_ParameterListInterpreterTpetra_MPI_1
        Start 1031: SEACASExodus_CreateEdgeFaceWithConcats_MPI_1
        Start 1983: ROL_example_PDE-OPT_helmholtz_example_02_MPI_1
        Start 501: TpetraCore_gemm_m_eq_5_MPI_1
  1/2160 Test #1031: SEACASExodus_CreateEdgeFaceWithConcats_MPI_1 ........................................................   Passed    0.73 sec
        Start 500: TpetraCore_gemm_m_eq_2_MPI_1
  2/2160 Test #1983: ROL_example_PDE-OPT_helmholtz_example_02_MPI_1 ......................................................   Passed  393.19 sec
        Start 499: TpetraCore_gemm_m_eq_1_MPI_1
  3/2160 Test #1621: MueLu_ParameterListInterpreterTpetra_MPI_1 ..........................................................***Timeout 600.10 sec
        Start 502: TpetraCore_gemm_m_eq_13_MPI_1
  4/2160 Test #501: TpetraCore_gemm_m_eq_5_MPI_1 ........................................................................***Timeout 600.12 sec
        Start 1370: Intrepid2_unit-test_Discretization_Basis_HierarchicalBases_Hierarchical_Basis_Tests_MPI_1
  5/2160 Test #500: TpetraCore_gemm_m_eq_2_MPI_1 ........................................................................***Timeout 600.06 sec
        Start 2142: PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1
...
  11/2160 Test #2142: PanzerAdaptersSTK_PoissonInterfaceExample_2d_diffsideids_MPI_1 ......................................   Passed  273.35 sec
```

**Comment by rppawlo** [comment]
Here's the gemm test output for the tests that were timing out above after deleting the test directory info:
```
        Start 499: TpetraCore_gemm_m_eq_1_MPI_1
439/2160 Test #481: TpetraTSQR_SequentialTsqr_contiguousCacheBlocks_MPI_1 ...............................................   Passed    3.53 sec
        Start 500: TpetraCore_gemm_m_eq_2_MPI_1
440/2160 Test #482: TpetraTSQR_SequentialTsqr_noncontiguousCacheBlocks_MPI_1 ............................................   Passed    3.54 sec
        Start 501: TpetraCore_gemm_m_eq_5_MPI_1
441/2160 Test #498: TpetraCore_gemv_MPI_1 ...............................................................................   Passed   19.72 sec
        Start 502: TpetraCore_gemm_m_eq_13_MPI_1
442/2160 Test #499: TpetraCore_gemm_m_eq_1_MPI_1 ........................................................................   Passed  199.91 sec
        Start 504: TpetraCore_BlockMultiVector2_MPI_1
443/2160 Test #501: TpetraCore_gemm_m_eq_5_MPI_1 ........................................................................   Passed  199.90 sec
        Start 507: TpetraCore_BlockView_MPI_1
444/2160 Test #500: TpetraCore_gemm_m_eq_2_MPI_1 ........................................................................   Passed  200.03 sec
```

**Comment by bartlettroscoe** [comment]
> On the second run through, the tests were reordered since ctest stored the runtimes and uses to optimize testing time. This second time through I got all the timeouts below. I think all the slow single process gpu intensive tests get launched together at the very beginning of a retest and jam the gpu.

@rppawlo, that is interesting and that makes sense.  That is the heuristic that CTest uses as it tries to reduce the total walllclock time.  But it does not understand the impact this can have on a GPU.

> I wonder if disabling the ctest optimization might help with this? Deleting the test output directory to wipe the timing info eliminates all the timeouts on a rerun.

That is an interesting idea.   I think the file you want to delete in the build directory is `Testing/Temporary/CTestCostData.txt`.   I can see about adding an option to `tribits_ctest_driver()` to delete that file if it exists before running the tests each time.  But this should only be added for CUDA builds as this should be an advantage to use that reordering algorithm for regular CPU cases.

But the fact that all of the timeouts are avoided from a fresh run is just dumb luck in our case.  If a single CMakeLists.txt file defined a bunch of expensive one-proc tests, would this problem not occur anyway?

**Comment by rppawlo** [comment]
I thought Matt put in an algorithm to pick random gpu device when there are multiples on a node. Could it be that all tests are using a single gpu? Maybe a flag needs to be set to tell testing how many gpus per node are present to sue Matt's code?

**Comment by bartlettroscoe** [comment]
> I thought Matt put in an algorithm to pick random gpu device when there are multiples on a node.

@rppawlo, where is that?  I think options like `KOKKOS_NUM_DEVICES` only work within a single MPI job and don't work across multiple MPI jobs.  Kitware extended ctest to target test jobs to different GPUs which was released in CMake 3.16 as described in:

* https://cmake.org/cmake/help/v3.16/manual/ctest.1.html#resource-allocation

but we can't use it yet set it requires support in Kokkos to read in the env vars that CTest sets described in:

* https://cmake.org/cmake/help/v3.16/manual/ctest.1.html#ctest-resource-environment-variables

Motivation and ideas on how this could apply to Trilinos are in:

* https://gitlab.kitware.com/snl/project-1/issues/100
* https://github.com/trilinos/Trilinos/pull/5598

Kitware just needs to get back to:

* https://gitlab.kitware.com/snl/project-1/issues/100

and work with the Kokkos developers to get this integrated with Kokkos and we can do what we need on the TriBITS and Trilinos side of things.

**Comment by mhoemmen** [comment]
@rrdrake has had success with environment variables like `KOKKOS_DEVICES`, but that's within a single job.

**Comment by rppawlo** [comment]
Matt's code was to target multiple separate executables in ctest: 

https://github.com/trilinos/Trilinos/pull/5799

https://github.com/kokkos/kokkos/issues/2254

**Comment by rppawlo** [comment]
That's a good idea! Would we be able to set this only for the cuda builds in the machine specific atdm scripts? I'd rather not force them to be serial on non-cuda platforms.

**Comment by bartlettroscoe** [comment]
> Would we be able to set this only for the cuda builds in the machine specific atdm scripts? 

Yes.  I will add this as a feature to TriBITS and then create a PR to mark the timing out tests in this issue and the issues #6804 and #6805.
